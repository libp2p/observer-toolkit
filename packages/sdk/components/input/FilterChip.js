import React, { useContext } from 'react'
import T from 'prop-types'
import styled from 'styled-components'
import { Formik } from 'formik'
import isEqual from 'lodash.isequal'

import Icon from '../Icon'
import Chip from '../Chip/Chip'
import Tooltip from '../Tooltip'
import { RootNodeContext } from '../context/RootNodeProvider'

const Container = styled.span`
  display: inline-block;
  margin: ${({ theme }) => theme.spacing()};
`

const RemoveIconButton = styled.a`
  &:hover {
    background: ${({ theme }) => theme.color('background', 0, 0.5)};
  }
`

const CheckIconButton = styled.a`
  border-radius: 50%;
  background: ${({ theme }) => theme.color('background', 0)};
  &:hover {
    background: ${({ theme }) => theme.color('background', 0, 0.5)};
  }
`

const Positioner = styled.div`
  left: ${({ theme }) => theme.spacing(-4)};
  transform: none;
`

const Tick = styled.div`
  left: ${({ theme }) => theme.spacing(6)};
  transform: none;
`

function FilterChip({
  filter: { name, enabled, values: filterValues, getFilterDef },
  dispatchFilters,
  side = 'bottom',
  isOpen = false,
  hidePrevious = null,
  format = null,
}) {
  const {
    FilterUi,
    initialValues,
    filterUiProps,
    valueNames = Object.keys(initialValues),
  } = getFilterDef()

  const rootNodeRef = useContext(RootNodeContext)

  const dispatch = (actionName, values) =>
    dispatchFilters({
      action: actionName,
      name,
      values,
    })

  const handleChange = values => dispatch('update', values)
  const disable = e => {
    e.stopPropagation()
    dispatch('disable')
  }
  const enable = e => {
    e.stopPropagation()
    dispatch('enable')
  }

  const chipOptions = {
    active: {
      colorKey: 'highlight',
    },
    inactive: {
      colorKey: 'text',
      colorIndex: 2,
    },
  }

  // If new values have been added, we need to merge them in
  const areValuesCurrent =
    filterValues &&
    isEqual(Object.keys(filterValues), Object.keys(initialValues))
  const formInitialValues = areValuesCurrent
    ? filterValues
    : Object.assign({}, initialValues, filterValues)

  return (
    <Formik
      initialValues={formInitialValues}
      onSubmit={(values, { setSubmitting }) => {
        handleChange(values)
        setSubmitting(false)
      }}
      enableReinitialize
    >
      {({ values, resetForm, setFieldValue, submitForm }) => {
        const reset = e => {
          e.stopPropagation()
          resetForm(initialValues)
          submitForm()
        }
        const chipType = enabled ? 'active' : 'inactive'
        const hasValues = !isEqual(values, initialValues)

        const chipPrefix = hasValues ? (
          <Tooltip
            side="top"
            fixOn="never"
            content={`${enabled ? 'Disable' : 'Re-enable'} filter`}
          >
            <Icon
              active={enabled}
              onClick={enabled ? disable : enable}
              type={enabled ? 'check' : 'uncheck'}
              override={{ Container: CheckIconButton }}
            />
          </Tooltip>
        ) : (
          ''
        )

        const chipSuffix = hasValues ? (
          <Tooltip side="top" fixOn="never" content="Reset filter">
            <Icon
              onClick={reset}
              type="remove"
              override={{ Container: RemoveIconButton }}
            />
          </Tooltip>
        ) : (
          ''
        )

        return (
          <Container>
            <Tooltip
              side={side}
              containerRef={rootNodeRef}
              fixOn={'no-hover'}
              initiallyOpen={isOpen}
              hidePrevious={hidePrevious}
              toleranceY={null}
              override={{ Positioner, Tick }}
              content={
                <FilterUi
                  onChange={submitForm}
                  values={values}
                  setFieldValue={setFieldValue}
                  fieldNames={valueNames}
                  title={name}
                  format={format}
                  {...filterUiProps}
                />
              }
            >
              <Chip
                type={chipType}
                options={chipOptions}
                prefix={chipPrefix}
                suffix={chipSuffix}
              >
                {name}
              </Chip>
            </Tooltip>
          </Container>
        )
      }}
    </Formik>
  )
}

FilterChip.propTypes = {
  filter: T.shape({
    name: T.string.isRequired,
    values: T.object,
    enabled: T.bool,
    getFilterDef: T.func.isRequired,
  }).isRequired,
  dispatchFilters: T.func.isRequired,
  side: T.string,
  isOpen: T.bool,
  hidePrevious: T.func,
  format: T.func,
  children: T.node,
}

export default FilterChip
