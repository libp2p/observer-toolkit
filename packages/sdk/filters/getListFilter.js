import CheckboxList from '../components/input/fields/CheckboxList'

const DefaultUi = CheckboxList

function getListFilter({
  valueNames,
  defaultValue = true,
  name = 'Filter from list',
  doFilter,
  FilterUi = DefaultUi,
  includeMissingValues = true,
  mapFilter,
}) {
  if (!valueNames || !Array.isArray(valueNames))
    throw new Error(
      `listFilter requires an array of value names, got ${typeof valueNames}`
    )

  const defaultFilter = (targetValue, values) => {
    if (targetValue === null || targetValue === undefined)
      return includeMissingValues
    return !!values[targetValue]
  }

  const initialValues = valueNames.reduce((values, name) => {
    values[name] = defaultValue
    return values
  }, {})

  return {
    name,
    doFilter: doFilter || defaultFilter,
    FilterUi,
    initialValues,
    valueNames,
    mapFilter,
  }
}

export default getListFilter
