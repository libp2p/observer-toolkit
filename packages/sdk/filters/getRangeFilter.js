import Slider from '../components/input/fields/Slider'

const defaultFilter = (targetValue, values) => {
  // If a value is null, that end of the rnage is unset, so allow all
  if (typeof values['min'] === 'number' && targetValue < values['min'])
    return false
  if (typeof values['max'] === 'number' && targetValue > values['max'])
    return false
  return true
}
const DefaultUi = Slider

function getRangeFilter({
  steps,
  min = 0,
  max = 0,
  name = 'Filter by range',
  doFilter = defaultFilter,
  FilterUi = DefaultUi,
}) {
  // Use empty strings for null values so form elements don't become uncontrolled
  const initialFieldValues = new Map([
    ['min', ''],
    ['max', ''],
  ])

  return {
    name,
    doFilter,
    FilterUi,
    initialFieldValues,
    filterUiProps: {
      min,
      max,
      steps,
    },
  }
}

export default getRangeFilter
