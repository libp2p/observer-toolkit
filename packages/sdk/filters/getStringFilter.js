import StringMatch from '../components/input/fields/StringMatch'

const defaultFilter = (targetValue, values) => {
  const matcher = values.regexp ? getRegExp(values.string) : values.string

  if (matcher instanceof Error) return true

  const match =
    matcher instanceof RegExp
      ? targetValue.match(matcher)
      : targetValue.includes(matcher)
  return values.exclude !== !!match
}
const DefaultUi = StringMatch

const initialValues = {
  regexp: false,
  exclude: false,
  string: '',
}

const valueNames = Object.keys(initialValues)

function getStringFilter({
  name = 'Filter by text',
  doFilter = defaultFilter,
  FilterUi = DefaultUi,
  mapFilter,
  ...inputProps
}) {
  return {
    name,
    doFilter,
    FilterUi,
    initialValues,
    valueNames,
    mapFilter,
    filterUiProps: {
      ...inputProps,
    },
  }
}

function getRegExp(str) {
  try {
    const regexp = new RegExp(str)
    return regexp
  } catch (err) {
    return err
  }
}

export default getStringFilter
