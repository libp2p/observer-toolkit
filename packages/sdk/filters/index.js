export { default as getListFilter } from './getListFilter'
export { default as getRangeFilter } from './getRangeFilter'
export { default as getStringFilter } from './getStringFilter'
