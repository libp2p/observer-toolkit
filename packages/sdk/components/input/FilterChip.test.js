import React, { useContext } from 'react'
import T from 'prop-types'
import { fireEvent } from '@testing-library/react'

import { renderWithTheme, within } from '@libp2p-observer/testing'

import {
  DataContext,
  DataProvider,
  FilterContext,
  FilterSetterContext,
  FilterProvider,
} from '../context'
import { getListFilter } from '../../filters'
import FilterChip from './FilterChip'

const mockAs = [
  { type: 'a', num: 2 },
  { type: 'a', num: 7 },
]
const mockBs = [
  { type: 'b', num: 4 },
  { type: 'b', num: 5 },
]
const mockCs = [
  { type: 'c', num: 6 },
  { type: 'c', num: 3 },
]
const mockData = [...mockAs, ...mockBs, ...mockCs]
mockData

const filterName = 'mock list filter'

const mockListFilter = getListFilter({
  name: filterName,
  mapFilter: datum => datum.type,
  // arbitrary non-alphabetical non-data order to ensure order is followed
  valueNames: ['b', 'a', 'c'],
})

function MockDataReceiver() {
  const data = useContext(DataContext)
  const { applyFilters } = useContext(FilterContext)
  const textContent = JSON.stringify(data.filter(applyFilters))
  return <div data-testid="output">{textContent}</div>
}

function FilterChipStack({ filterDef }) {
  return (
    <FilterProvider filterDefs={[filterDef]}>
      <DataProvider initialData={{ states: mockData }}>
        <FilterChipWrapper />
        <MockDataReceiver />
      </DataProvider>
    </FilterProvider>
  )
}
FilterChipStack.propTypes = {
  filterDef: T.object.isRequired,
}

function FilterChipWrapper() {
  const dispatchFilters = useContext(FilterSetterContext)
  const { filters } = useContext(FilterContext)
  return <FilterChip filter={filters[0]} dispatchFilters={dispatchFilters} />
}

describe('FilterChip', () => {
  it('renders, matching snapshot', () => {
    const { asFragment } = renderWithTheme(
      <FilterChipStack filterDef={mockListFilter} />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('opens working list filter UI on click', async () => {
    const {
      findByText,
      getByRole,
      getByText,
      queryAllByRole,
    } = renderWithTheme(<FilterChipStack filterDef={mockListFilter} />)

    // No checkboxes, no filtering before clicking on chip
    expect(queryAllByRole('checkbox')).toHaveLength(0)
    expect(await findByText(JSON.stringify(mockData))).toBeTruthy()

    const chip = getByText(filterName)
    await fireEvent.click(chip)

    // Checkboxes are unmounted and created new when dialog opens / closes
    const queryCheckboxes = () =>
      within(getByRole('tooltip')).queryAllByRole('checkbox')
    const initialCheckboxes = queryCheckboxes()
    expect(initialCheckboxes).toHaveLength(4)

    // First ordinary checkbox should filter "b"s according to our b, a, c order
    await fireEvent.click(initialCheckboxes[1])
    expect(
      await findByText(JSON.stringify([...mockAs, ...mockCs]))
    ).toBeTruthy()

    // Second ordinary checkbox should filter "a"s as well as "b"s
    await fireEvent.click(initialCheckboxes[2])
    expect(await findByText(JSON.stringify(mockCs))).toBeTruthy()

    // Opening and closing the panel should not change what is filtered
    await fireEvent.click(chip)
    expect(queryAllByRole('tooltip')).toHaveLength(0)
    await fireEvent.click(chip)
    const secondCheckboxes = queryCheckboxes()
    expect(secondCheckboxes).toHaveLength(4)
    expect(await findByText(JSON.stringify(mockCs))).toBeTruthy()

    // Third ordinary checkbox should filter "c"s leaving nothing
    await fireEvent.click(secondCheckboxes[3])
    expect(await findByText(JSON.stringify([]))).toBeTruthy()
  })

  it('filters all when toggle-all button used', async () => {
    const { findByText, getByRole, getByText } = renderWithTheme(
      <FilterChipStack filterDef={mockListFilter} />
    )
    await fireEvent.click(getByText(filterName))
    const checkboxes = within(getByRole('tooltip')).getAllByRole('checkbox')
    expect(checkboxes[0]).toBeChecked()
    expect(checkboxes[1]).toBeChecked()
    expect(checkboxes[2]).toBeChecked()

    // First checkbox should filter everything
    await fireEvent.click(checkboxes[0])
    expect(await findByText(JSON.stringify([]))).toBeTruthy()
    expect(checkboxes[0]).not.toBeChecked()
    expect(checkboxes[1]).not.toBeChecked()
    expect(checkboxes[2]).not.toBeChecked()

    // Clicking an ordinary checkbox after filtering everything shows not hides
    await fireEvent.click(checkboxes[1])
    expect(await findByText(JSON.stringify(mockBs))).toBeTruthy()
    expect(checkboxes[0]).toHaveAttribute('aria-checked', 'mixed')
    expect(checkboxes[1]).toBeChecked()
    expect(checkboxes[2]).not.toBeChecked()

    // Clicking 'indeterminate' / 'mixed' toggle-all shows all
    await fireEvent.click(checkboxes[0])
    expect(await findByText(JSON.stringify(mockData))).toBeTruthy()
    expect(checkboxes[0]).toBeChecked()
    expect(checkboxes[1]).toBeChecked()
    expect(checkboxes[2]).toBeChecked()
  })

  it('disables but saves filters on disable, wipes on clear', async () => {
    const { findByText, getByRole, getByText } = renderWithTheme(
      <FilterChipStack filterDef={mockListFilter} />
    )
    await fireEvent.click(getByText(filterName))
    const checkboxes = within(getByRole('tooltip')).getAllByRole('checkbox')
    expect(checkboxes[0]).toBeChecked()
    expect(checkboxes[1]).toBeChecked()
    expect(checkboxes[2]).toBeChecked()

    // First checkbox should filter everything
    await fireEvent.click(checkboxes[0])
    expect(await findByText(JSON.stringify([]))).toBeTruthy()
    expect(checkboxes[0]).not.toBeChecked()
    expect(checkboxes[1]).not.toBeChecked()
    expect(checkboxes[2]).not.toBeChecked()

    // Clicking an ordinary checkbox after filtering everything shows not hides
    await fireEvent.click(checkboxes[1])
    expect(await findByText(JSON.stringify(mockBs))).toBeTruthy()
    expect(checkboxes[0]).toHaveAttribute('aria-checked', 'mixed')
    expect(checkboxes[1]).toBeChecked()
    expect(checkboxes[2]).not.toBeChecked()

    // Clicking 'indeterminate' / 'mixed' toggle-all shows all
    await fireEvent.click(checkboxes[0])
    expect(await findByText(JSON.stringify(mockData))).toBeTruthy()
    expect(checkboxes[0]).toBeChecked()
    expect(checkboxes[1]).toBeChecked()
    expect(checkboxes[2]).toBeChecked()
  })
})
