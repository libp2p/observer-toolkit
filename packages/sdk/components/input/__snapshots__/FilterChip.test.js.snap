// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterChip renders, matching snapshot 1`] = `
<DocumentFragment>
  .c2 {
  position: relative;
  display: inline-block;
  cursor: pointer;
}

.c0 {
  width: inherit;
  height: inherit;
  display: inherit;
  -webkit-flex-direction: inherit;
  -ms-flex-direction: inherit;
  flex-direction: inherit;
}

.c3 {
  display: inline-block;
  padding: 0px 4px;
  background: rgba(128,130,133,0.15);
  color: rgba(128,130,133,1);
  white-space: nowrap;
  font-size: 8pt;
  font-weight: 600;
}

.c4 {
  line-height: 2em;
  display: inline-block;
  padding: 4px;
  vertical-align: middle;
}

.c1 {
  display: inline-block;
  margin: 8px;
}

<div
    class="c0"
  >
    <span
      class="c1"
    >
      <span
        class="c2"
        data-tooltip="closed"
      >
        <span
          class="c3"
          style="box-shadow: 0 0 16px 0 rgb(128, 130, 133);"
        >
          <span
            class="c4"
          >
            mock list filter
          </span>
        </span>
      </span>
    </span>
    <div
      data-testid="output"
    >
      [{"type":"a","num":2},{"type":"a","num":7},{"type":"b","num":4},{"type":"b","num":5},{"type":"c","num":6},{"type":"c","num":3}]
    </div>
  </div>
</DocumentFragment>
`;
